### ISSUE TITLE
Support minmax reductions?

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#14770]
We have min and max reductions, but no way to get both at once. You can get the min and max separately but you end up looking through your array twice. You can optimize this by manually writing the loop yourself, but it's not as clean/simple:

```chpl
var A: [1..10] int = 1..10;

// Clean-ish, but goes through the array twice
{
  var aMin = min reduce A;
  var aMax = max reduce A;
  writeln((aMin, aMax));
}

// Uglier, but faster
{
  // I usually incorrectly write `var aMin, aMax: int`, which leads to subtle bugs
  var aMin = max(int); 
  var aMax = min(int);
  forall a in A with (min reduce aMin, max reduce aMax) {
    aMin reduce= a;
    aMax reduce= a;
  }
  writeln((aMin, aMax));
}
```

It seems attractive to be able to do something like:

```chpl
var (aMin, aMax) = minmax reduce A;
```

Finding both min and max is a pretty common idiom idiom in arkouda. Reductions are fast enough that I don't think it's a major bottleneck, but it would save some time.

**Comment by bradcray** [comment]
This would be trivial to do.  I marked it as such.

**Comment by bradcray** [comment]
(it also could beg the question whether we'd want a `minmaxlocs` reduction that was like minloc and maxloc combined).

**Comment by masterashu** [comment]
I would like to work on it.

**Comment by bradcray** [comment]
To get you started, I'd suggest looking at modules/internal/ChapelReduce.chpl, look at how the min and MinReduceOp and MaxReduceOp classes are defined, and add a similar class named `minmax`.  Doing so _ought_ to permit you to write `minmax reduce ...`  test/thomasvandoren/mink.chpl (and the nearby tests that use it) may also be a good reference point.  Sorry there's not better documentation on writing these operations, but it shouldn't be too hard to pattern-match against what others have done and get it working, I believe.  Ask questions if not.

**Comment by rahulghangas** [comment]
@masterashu I've got a working solution that I would like to make a pull request for. But I would understand if you object since you were assigned to this.

**Comment by masterashu** [comment]
@rahulghangas No that is fine.

**Comment by klakho0400** [comment]
is anyone working on this or can i take this issue

**Comment by rahulghangas** [comment]
I've got a branch (referenced above) that I have been meaning to make a create a pull request for. I still have to write tests for it, which I haven't gotten around to

**Comment by klakho0400** [comment]
oh ok

**Comment by bradcray** [comment]
@ronawho: PR #15062 implements `minmax` reductions.  This made me go looking for places in Arkouda where they could be used, and I'm finding cases in:

* https://github.com/mhmerrill/arkouda/blob/master/src/ArgSortMsg.chpl
* https://github.com/mhmerrill/arkouda/blob/master/src/AryUtil.chpl
* https://github.com/mhmerrill/arkouda/blob/master/src/FindSegmentsMsg.chpl
* https://github.com/mhmerrill/arkouda/blob/master/src/HistogramMsg.chpl
* https://github.com/mhmerrill/arkouda/blob/master/src/IndexingMsg.chpl
* https://github.com/mhmerrill/arkouda/blob/master/src/RadixSortLSD.chpl
* https://github.com/mhmerrill/arkouda/blob/master/src/ReductionMsg.chpl
* https://github.com/mhmerrill/arkouda/blob/master/src/SegmentedArray.chpl

(though there may be others as well).

This raises the perennial question about whether we should open a PR to be merged after the next release, or drop in a helper file that implements the minmax reduction as a placeholder, or file an issue reminding ourselves to do this change after the next release goes out).

**Comment by ronawho** [comment]
I don't think minmax is a big bottleneck, but I'd still be in favor of getting that into arkouda. Can we just drop in a duplicate minmax class into arkouda and have that take precedence over the builtin one that will be on master, or will that be ambiguous? If it's ambiguous I'd probably just name it `akminmax` with a comment saying "can use builtin minmax once 1.23 is required"

**Comment by vasslitvinov** [comment]
Resolved by #15062

**Comment by bradcray** [comment]
> Can we just drop in a duplicate minmax class into arkouda and have that take precedence over the builtin one that will be on master, or will that be ambiguous?

I believe if it's put into an Arkouda file that's explicitly `use`d, this should work.  But calling it `akminmax` may result in less confusion while achieving the same result.
